tasks/Task.cpp

```diff
@@ -2,48 +2,496 @@
+                dataPacket.data = new_data;
+            }
+            else
+            {
+                vector<byte> data_byte = get<binary>(data);
+                dataPacket.data.resize(data_byte.size());
+                for (unsigned int i = 0; i < data_byte.size(); i++)
+                {
+                    dataPacket.data[i] = to_integer<uint8_t>(data_byte[i]);
+                }
+            }
+            _data_out.write(dataPacket);
+        });
+}
+
+void Task::evaluateDataChannel()
 {
+    switch (mState.data_channel)
+    {
+    case DataChannelOpened:
+    {
+        iodrivers_base::RawPacket raw_packet;
+        if (_data_in.read(raw_packet) != RTT::NewData)
+            return;
+
+        vector<byte> data;
+        data.resize(raw_packet.data.size());
+        for (unsigned int i = 0; i < raw_packet.data.size(); i++)
+        {
+            data[i] = static_cast<byte>(raw_packet.data[i]);
+        }
+        mDataChannel->send(&data.front(), data.size());
+        break;
+    }
+    case DataChannelClosed:
+        throw std::runtime_error("DataChannel closed");
+    case DataChannelFailed:
+        throw std::runtime_error("DataChannel failed");
+    default:
+        break;
+    }
 }
 
-Task::~Task()
+void Task::evaluateWebSocket()
 {
+    switch (mState.web_socket)
+    {
+    case WebSocketClosed:
+        throw std::runtime_error("WebSocket closed");
+    case WebSocketFailed:
+        throw std::runtime_error("WebSocket failed");
+    default:
+        break;
+    }
 }
 
+Task::Task(string const &name) : TaskBase(name) {}
 
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+    mConfig.iceServers.emplace_back(_stun_server.get());
+    mWs = std::make_shared<rtc::WebSocket>();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
+
+    if (!_remote_peer_id.get().empty())
+    {
+        std::future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+        mRemotePeerID = _remote_peer_id.get();
+        mRemotePeerAnswerReceived = false;
+
+        base::Time start_time = base::Time::now();
+        // Get contact with the remote peer and create datachannel
+        while (!mRemotePeerAnswerReceived)
+        {
+            base::Time duration = base::Time::now() - start_time;
+            if (duration > _wait_remote_peer_time_out.get())
+            {
+                mWaitRemotePeerPromise.set_exception(
+                    std::make_exception_ptr(std::runtime_error("Timeout to find the remote peer")));
+                break;
+            }
+            ping();
+            sleep_for(100ms);
+        }
```

Comment:
There is no future in this case, just the ping and mRemotePeerAnswerReceived

Reply:
And I'd suggest moving the whole block of code into a 'waitRemotePeerPresent' method or something.

Reply:
> There is no future in this case, just the ping and mRemotePeerAnswerReceived
I didn't understand....and I'm still thinking how to use `wait_for` or `wait_until`methods in that situation, because will block the thread (so it is not possible to use inside the loop, and don't make sense to use outside the loop beause if we are at this point, we have a `set_value`)